rfd/0127-encrypted-session-recordings.md

```diff
@@ -0,0 +1,489 @@
+authors: Alan Parra (alan.parra@goteleport.com), Erik Tate (erik.tate@goteleport.com) 
+state: draft
+---
+
+# RFD 127 - Encrypted Session Recordings
+
+## Required Approvers
+* Engineering: @rosstimothy, @zmb3, @espadolini, @nklaassen
+* Security: doyensec
+
+## What
+
+This document proposes an approach to encrypting session recording data before
+writing to disk or any long term storage.
+
+## Why
+
+Recordings temporarily stored to disk can be easily tampered with by users with
+enough access. This could even occur within the session being recorded if the
+host user has root access.
+
+Encrypting session recordings at rest can help prevent exposure of credentials
+or other secrets that might be visible within the recording. 
+
+## Details
+
+This document should fulfill the following requirements:
+- Ability to encrypt session recording data at rest in long term storage and
+during any intermediate disk writes.
+- Ability to replay encrypted sessions using the web UI.
+- Ability to guard decryption using key material from an HSM or other supported
+  keystore.
+- Support for multiple auth servers using different HSM/KMS backends.
+- An encryption algorithm suitable for this workload.
+
+### Encryption Algorithm
+
+This RFD assumes the usage of [age](https://github.com/FiloSottile/age), which
+was chosen for its provenance, simplicity, and focus on strong cryptography
+defaults without requiring customization. The formal spec can be found
+[here](https://age-encryption.org/v1). Officially supported key algorithms are
+limited to X25519 (recommended by the spec), Ed25519, and RSA. Support for
+other algorithms would either have to be requested from the upstream or
+manually implemented as a custom plugin. The algorithms employed by `age` are
+not currently compatible with FIPS, which means configuring encrypted sessions
+while in FIPS mode will result in failed startup.
+
+Below is a high level diagram showing how `age` encryption and decryption work:
+![age diagram](assets/0127-age-high-level.png)
+
+### Config Changes
+
+Encrypted session recording is a feature of the auth service and can be enabled
+through the `session_recording_config` resource.
+```yaml
+# session_recording_config.yml
+kind: session_recording_config
+version: v2
+spec:
+  encryption:
+    enabled: true
+```
+HSM integration is facilitated through the existing configuration
+options for setting up an HSM backed CA keystore through pkcs#11. Example
+configuration found [here](https://goteleport.com/docs/admin-guides/deploy-a-cluster/hsm/#step-25-configure-teleport).
+
+### Protobuf Changes
+```proto
+// api/proto/teleport/legacy/types/types.proto
+
+// EncryptionKeyPair is a keypair used for encrypting and decrypting data.
+message EncryptionKeyPair {
+  // PublicKey is the SSH public key.
+  bytes PublicKey = 1 [(gogoproto.jsontag) = "public_key,omitempty"];
+  // PrivateKey is the SSH private key.
+  bytes PrivateKey = 2 [(gogoproto.jsontag) = "private_key,omitempty"];
```

Comment:
the mention of SSH here is a bit of a red herring. Will these be serialized in OpenSSH format, except for HSM keys, which will follow the existing HSM key format I presume?

Reply:
I accidentally left `SSH` in these comments from the previous proto definition. In practice these will either be `X25519` keys stored as `Raw`, or HSM/KMS keys which will follow the existing key format